File: patterns/rate-limit/rules-python/rate_limit.py
from datetime import datetime, timedelta

from openhab import rule
from openhab.triggers import ItemCommandTrigger
from scope import cache


@rule(triggers=[ItemCommandTrigger("RateLimitTrigger", "ON")])
class RateLimitBeispiel:
    def execute(self, module, input):
        last_action = cache.privateCache.get("LastAction")
        now = datetime.now().astimezone()

        if last_action is None or last_action < now - timedelta(hours=24):
            self.logger.info("Rate-limited action")
            cache.privateCache.put("LastAction", now)
        else:
            self.logger.info("Ereignis ignoriert, Sperrzeit laeuft noch")

File: patterns/rate-limit/rules-js/rate-limit.js
rules.JSRule({
  name: "Rate-Limit Beispiel",
  triggers: [triggers.ItemCommandTrigger("RateLimitTrigger", "ON")],
  execute: (event) => {
    const lastAction = cache.private.get("LastAction", () => time.ZonedDateTime.now().minusDays(1));

    if (lastAction.isBefore(time.ZonedDateTime.now().minusHours(24))) {
      console.log("Rate-limited action");
      cache.private.put("LastAction", time.ZonedDateTime.now());
    } else {
      console.log("Ereignis ignoriert, Sperrzeit laeuft noch");
    }
  }
});
